### ISSUE TITLE
libgearman isn't hashing submitted jobs based on unique

### ISSUE THREAD

**Issue by SpamapS** [gearman/gearmand#64]
In the original gearman client, the client would take the list of servers it knows about, hash the unique ID of each submitted job, and choose one of the servers in the list based on the hash of the unique.

This code seems to have been lost, and no tests enforce it. This is an important part of what makes gearman awesome, so it needs to be resurrected.

**Comment by sirgalleto** [comment]
@SpamapS I actually have a bug with the queue creation with mysql because the queue is trying to create a duplicate entry in the database. Please tell me if we are talking about the same bug and if you found any kind of solution.

Thanks

**Comment by p-alik** [comment]
@sirgalleto, see mein comment in the isseu #80

**Comment by SpamapS** [comment]
Indeed, @sirgalleto this is not related. This is about clients of many gearmand's that are selecting the wrong gearmand to submit their job to.

**Comment by p-alik** [comment]
> In the original gearman client, the client would take the list of servers it knows about, hash the unique ID of each submitted job, and choose one of the servers in the list based on the hash of the unique.

@SpamapS, could you point me to the code part you mentioned here?

**Comment by bmeynell** [comment]
@p-alik - This issue appears to have been discovered [here](https://groups.google.com/d/msg/gearman/iAE-3TYSd0w/51nT6R9LBgAJ) and first logged [here](https://bugs.launchpad.net/gearmand/+bug/1481993). This is a huuuuuge gaping hole preventing gearmand from being a true "distributed nervous system". I would also be keen to see the portion(s) of the C code that would need modifications if anyone has that on their radar.

**Comment by p-alik** [comment]
Dug into the current code to answer question about where the relevant code lives. Confirming the bug is real and still present.

There's no server-selection hashing anywhere in libgearman today. Two unrelated pieces of infrastructure make it *look* like there might be:

1. `gearman_unique_is_hash()` (`libgearman/unique.cc:57`) + `add_task()`/`add_reducer_task()` (`libgearman/add.cc:213`, `add.cc:423`): if you pass `unique = "#"`, the client derives the job's `unique` value from `murmur3(workload)`. Documented in `ChangeLog:87`. This generates a *unique ID* — it has nothing to do with picking a server.

2. `libhashkit` is bundled and linked into libgearman (`libgearman/include.am:155`). It's the same consistent-hashing library libmemcached uses, complete with distribution-function support (`hashkit_set_distribution_function`, etc. in `libhashkit-1.0/function.h`). But it's only ever called for the murmur3 workload hash above — it's never wired up to pick a connection.

Actual server selection happens in `libgearman/run.cc:74-81` (and the retry path at `run.cc:120-128`):

```cpp
for (task->con= task->client->universal.con_list; task->con;
     task->con= task->con->next_connection())
{
  if (task->con->send_state == GEARMAN_CON_SEND_STATE_NONE)
  {
    break;
  }
}
```

This walks `con_list` from the head and takes the first connection that isn't mid-send. It never looks at `task->unique`. And per `connection.cc:119-124`, new connections are *prepended* to `con_list`, so the list is in reverse order of `gearman_client_add_server()` calls — meaning in practice jobs stick to whichever server was added *last*, until that connection is busy sending, then fail over to the next-most-recently-added, and so on. It's not round-robin and it's not hash-based; it's "most-recently-added-and-free."

So @SpamapS's description of the original client behavior (hash unique → pick server from list, memcached-style) does not exist in this codebase. The plumbing for real consistent hashing (`libhashkit`) is already a dependency and sitting unused for this purpose — a fix would presumably wire `hashkit_set_distribution_function`/`libhashkit_generate` into the connection-selection loop in `run.cc`, keyed on `task->unique`, instead of the current head-of-list scan.

**Comment by p-alik** [comment]
Before implementing, this needs a few design decisions — the root cause (`run.cc:74-81` ignoring `task->unique`) is clear, but "what should happen instead" has real options with different tradeoffs:

**1. Which `libhashkit` distribution function to key on**
- **Modula** (`HASHKIT_DISTRIBUTION_MODULA`) — simplest, matches classic memcached-client hashing (`hash(unique) % server_count`). Downside: adding/removing a server remaps almost every job to a different server.
- **Ketama / consistent hashing** (`HASHKIT_DISTRIBUTION_CONSISTENT_KETAMA`) — minimizes remapping when the server list changes, which matters more for a long-lived worker/client fleet than for a cache. Slightly more code (`libhashkit` already supports it, just unwired).
- **Match whatever the original C client actually did** — nobody in this thread has pointed at the original source, only a mailing-list post and a Launchpad bug report describing the symptom. If someone has the pre-regression client code or version, that settles this outright.

**2. Failover when the hash-selected server is busy or down**
- **Strict**: only ever use the hash-selected server; if it's unavailable, the job submission fails (matches "hash picks the server" literally, but removes the current head-of-list failover behavior entirely).
- **Ordered fallback**: hash picks the primary, but on failure fall through to the next connection in hash-ring order (ketama gives this naturally; modula would need an explicit secondary rule).
- **Keep current fallback, just fix the primary pick**: hash-select first, then fall back to today's "first free connection in `con_list`" scan if the hash pick is unavailable — smallest behavior change, but mixes two selection strategies.

**3. Scope / rollout**
- **Change the default** for all multi-server clients — matches "restore old behavior," but is a silent behavior change for anyone currently relying on (or accidentally depending on) the current most-recently-added-and-free scan.
- **Opt-in via a new API/option** (e.g. `gearman_client_set_server_selection(...)`) — safer for existing deployments, but means the "broken" default stays broken unless callers opt in.

Given the compatibility angle, this is why #460 flagged 2.2.0 rather than a patch release — happy to start on this once (1)-(3) are settled, or take a stab at a specific combination if there's a preference.
